design_notes/v2.md

```diff
@@ -0,0 +1,98 @@
+# VirtualiZarr V2 architecture discussion
+
+## Goals
+
+- Extensible and easily understandable entry-point system for new backends
+- Modular code base that uses Obspec for the object storage interface abstraction, Zarr as the core data model, and Xarray for the concatenation and merging of Zarr stores
+
+## Related issues
+
+- [Fragility of url auto-parsing logic](https://github.com/zarr-developers/VirtualiZarr/issues/561)
+- [Use obstore / obspec for globbing in `open_virtual_mfdataset`](https://github.com/zarr-developers/VirtualiZarr/issues/569)
+- [Function definition for open_virtual_dataset in next major release](https://github.com/zarr-developers/VirtualiZarr/issues/553)
+- [Support passing configuration options to default_object_store](https://github.com/zarr-developers/VirtualiZarr/issues/559)
+- [Redefine Virtual Readers as func(filepath) -> ManifestStore](https://github.com/zarr-developers/VirtualiZarr/issues/498)
+- [Make xarray an optional dependency?](https://github.com/zarr-developers/VirtualiZarr/issues/521)
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: An object defining the paths, chunk byte offsets, lengths of one virtual array.
+- `ManifestArray`: A chunk manifest and its associated Zarr V3 ArrayMetadata
+- `ManifestGroup`: ManifestGroups are a way to organizing child groups and arrays into a data hierarchy. Contains Zarr `GroupMetadata` and one or more ManifestGroups and/or ManifestArrays.
+- `ManifestStore`: A ManifestGroup and its associated ObjectReader (see protocol below).
+- `xarray.Dataset`: Contains ManifestArray-backed variables and (optionally) in-memory NumPy-backed variables.
+
+## Definitions
+
+- `ObjectReader` - An implementation of the `GetAsync` and `GetRangeAsync` protocols defined in obspec. `ObjectReader`s are used to load arrays using `ManifestStore`.
```

Comment:
Does this also need `List` support?

And this says that it requires `GetAsync` and `GetRangeAsync`, but that implies that 1) all backends need to be async-capable and that 2) the `ObjectReader` itself will be async (or internally have an event loop that blocks on all the async calls)

It's annoying but unavoidable that async makes you deal with [colored functions](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

You should decide whether the required performance characteristics of Virtualizarr require exposing an async API to the user. If not, then maybe only use blocking calls here (`Get` and `GetRange` and `GetRanges`). If so, then you might need an `ObjectReader` and `AsyncObjectReader`.

Reply:
The async get range method is primarily used within `ManifestStore`'s implementation of Zarr's async Store API. So I think we definitely want an async version - https://github.com/zarr-developers/VirtualiZarr/blob/5646b675e37ded8d18037bf0aab9a0786650c46c/virtualizarr/manifests/store.py#L322-L328

(note I should change this to use `store.get_range_async` to allow interoperability following obspec).

I'll revise to add an `ObjectReader` and `AsyncObjectReader` with the `AsyncObjectReader` required by `ManifestStore` and the `ObjectReader` available for use in backends.